ultra/docs/curriculum_learning.md

```diff
@@ -0,0 +1,126 @@
+
+You might wonder that ULTRA itself is defined to be a single task problem, solving the unprotected left-turn. However, agents in ULTRA learn from scratch which means gaining experiences in the basics of driving, centering in lanes, stopping behind other vehicles and on different types of traffic density. Then after we can focus on the target task (unprotected left-turn). Instead of presenting randomly selected scenarios to an agent, we can construct a curriculum which prioritizes the agent to first pass the easier scenarios.
+
+Curricula can be either created before hand or on-the-fly. In ULTRA, we refer to curricula which are built prior to training as "static" and curricula that are not fully established or can change during training are referred as "dynamic".
+
+**Static curricula** in ULTRA currently use an implementation called the grade-based (GB) structure. The grade-based structure allows scenarios to be sequenced/ordered during training and testing runs. This is done by dividing an experiment into so called "grades", which are combinations of different tasks and levels. These grades are basically buckets for combinations of tasks and their levels.
+
+When an agent is put into training with curriculum-mode enabled (`ultra/train.py` is run with flag `--curriculum-mode True`), a coordinator object is initialized. It is responsible for setting up the scenarios in each grade and performing "graduation": switching of the agent into the next grade.
+
+Unlike the regular training setup (when curriculum mode is disabled), there is another task folder called `curriculum/` under `ultra/scenarios/` which has the blueprint of which set of scenarios are in each grade:
+
+```yaml
+curriculum:
+  static:
+    toggle: <bool> # Turn on static curriculum settings
+    grades:
+      1: [[<task>,<level>]]
+      2: [[<task>,<level>], [<task>,<level>]]
+      # 3: [[<task>,<level>], ... ,[<task>,<level>]] can have as many combinations of tasks and levels
+    conditions:
+      eval_per_grade: <bool> # Evaluation takes places after completion of grade. An "exam" at the end of year
+      episode_based: # Agent graduates after completing a N number of episodes in each grade
+        toggle: <bool> # Enable the condition
+        cycle: <bool> # Option to keep cycling through grades to episodes limit
+      pass_based: # Agent graduates after getting an average completion rate, the average is taken over the eval-rate (sampling-rate)
+        toggle: <bool> # Enable the condition
```

Comment:
Just wondering, what happens if they are both toggled? Does the last one go-through, in this case pass_based? Also, could you clarify cycle: would you go grade 1 -> grade 2 -> grade 1 and so on?

Reply:
Oh yes, when the two conditions are toggled, then there is an warning that is issued at runtime.
https://github.com/huawei-noah/SMARTS/blob/ecd25bbf5fc853dddb991b40766bd8c0fb2ed642/ultra/ultra/utils/curriculum/curriculum_info.py#L80-L87

If the cycle flag is true then the grades will start back from the beginning, in  this case it will be grade1.

I will clarify those two things in the documentation